File: others/分类/双指针/42接雨水.cpp
/*
*/


#include <algorithm>
#include <vector>
#include <math.h>

using namespace std;

class Solution {
public:
    int trap(vector<int> &height) {
        //1
                // int ans = 0;
                // int left = 0, right = height.size() - 1;
                // int leftMax = 0, rightMax = 0;
                // while (left < right) {
                //     leftMax = max(leftMax, height[left]);
                //     rightMax = max(rightMax, height[right]);
                //     if (height[left] < height[right]) {
                //         ans += leftMax - height[left];
                //         ++left;
                //     } else {
                //         ans += rightMax - height[right];
                //         --right;
                //     }
                // }
                // return ans;

        //2
                //     int ans = 0;
            //     vector<int> leftMax(height.size());
            //     vector<int> rightMax(height.size());

            //     leftMax[0] = height[0];
            //     for (int i = 1; i < height.size(); ++i) {
            //         leftMax[i] = max(leftMax[i - 1], height[i]);
            //     }

            //     rightMax[0] = height[height.size() - 1];
            //     for (int i = height.size() - 2; i >= 0; --i) {
            //         rightMax[i] = max(rightMax[i + 1], height[i]);
            //     }

            //     for (int i = 1; i < height.size(); ++i) {
            //         ans += min(leftMax[i], rightMax[i]) - height[i];
            //     }

            //     return ans;



        int ans = 0;
        int left = 0, right = height.size() - 1;
        int leftMax = 0, rightMax = 0;
        while (left < right) {
            leftMax = max(leftMax, height[left]);
            rightMax = max(rightMax, height[right]);
            if (leftMax < rightMax) {
                ans += leftMax - height[left];
                ++left;
            } else {
                ans += rightMax - height[right];
                --right;
            }
        }
        return ans;
    }
    //3


};

File: others/0-100/42. 接雨水.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
MOD = int(1e9 + 7)


class Solution:
    def trap(self, height: List[int]) -> int:
        dq = deque()
        ans = 0
        for i, x in enumerate(height):
            while dq and x >= height[dq[-1]]:
                y = dq.pop()
                if dq:
                    ans += (i - dq[-1] - 1) * (min(height[dq[-1]], x) - height[y])
            dq.append(i)
        return ans


class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        right = [0] * n
        right[-1] = height[-1]
        for i in range(n - 2, -1, -1):
            right[i] = max(right[i + 1], height[i])
        left = 0
        ans = 0
        for i, x in enumerate(height):
            mn = min(left, right[i])
            if mn > x:
                ans += mn - x
            if x > left:
                left = x
        return ans
